// @trezor/transport/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UdpTransport = exports.SessionsClient = exports.SessionsBackground = exports.NodeUsbTransport = exports.WebUsbTransport = exports.BridgeTransport = exports.Messages = exports.UsbApi = exports.AbstractApiTransport = exports.isTransportInstance = exports.Transport = exports.TRANSPORT = exports.TREZOR_USB_DESCRIPTORS = exports.TRANSPORT_ERROR = void 0;
const tslib_1 = require("tslib");
const protobuf = tslib_1.__importStar(require("protobufjs/light"));
const long_1 = tslib_1.__importDefault(require("long"));
exports.TRANSPORT_ERROR = tslib_1.__importStar(require("./errors"));
protobuf.util.Long = long_1.default;
protobuf.configure();
var constants_1 = require("./constants");
Object.defineProperty(exports, "TREZOR_USB_DESCRIPTORS", { enumerable: true, get: function () { return constants_1.TREZOR_USB_DESCRIPTORS; } });
Object.defineProperty(exports, "TRANSPORT", { enumerable: true, get: function () { return constants_1.TRANSPORT; } });
var abstract_1 = require("./transports/abstract");
Object.defineProperty(exports, "Transport", { enumerable: true, get: function () { return abstract_1.AbstractTransport; } });
Object.defineProperty(exports, "isTransportInstance", { enumerable: true, get: function () { return abstract_1.isTransportInstance; } });
var abstractApi_1 = require("./transports/abstractApi");
Object.defineProperty(exports, "AbstractApiTransport", { enumerable: true, get: function () { return abstractApi_1.AbstractApiTransport; } });
var usb_1 = require("./api/usb");
Object.defineProperty(exports, "UsbApi", { enumerable: true, get: function () { return usb_1.UsbApi; } });
var protobuf_1 = require("@trezor/protobuf");
Object.defineProperty(exports, "Messages", { enumerable: true, get: function () { return protobuf_1.Messages; } });
var bridge_1 = require("./transports/bridge");
Object.defineProperty(exports, "BridgeTransport", { enumerable: true, get: function () { return bridge_1.BridgeTransport; } });
var webusb_1 = require("./transports/webusb");
Object.defineProperty(exports, "WebUsbTransport", { enumerable: true, get: function () { return webusb_1.WebUsbTransport; } });
var nodeusb_1 = require("./transports/nodeusb");
Object.defineProperty(exports, "NodeUsbTransport", { enumerable: true, get: function () { return nodeusb_1.NodeUsbTransport; } });
var background_1 = require("./sessions/background");
Object.defineProperty(exports, "SessionsBackground", { enumerable: true, get: function () { return background_1.SessionsBackground; } });
var client_1 = require("./sessions/client");
Object.defineProperty(exports, "SessionsClient", { enumerable: true, get: function () { return client_1.SessionsClient; } });
var udp_1 = require("./transports/udp");
Object.defineProperty(exports, "UdpTransport", { enumerable: true, get: function () { return udp_1.UdpTransport; } });
