// @trezor/blockchain-link-types/lib/constants/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TERMINATE = exports.PUSH_TRANSACTION = exports.UNSUBSCRIBE = exports.SUBSCRIBE = exports.RPC_CALL = exports.ESTIMATE_FEE = exports.GET_TRANSACTION_HEX = exports.GET_TRANSACTION = exports.GET_ACCOUNT_UTXO = exports.GET_ACCOUNT_INFO = exports.GET_BLOCK = exports.GET_BLOCK_HASH = exports.GET_FIAT_RATES_TICKERS_LIST = exports.GET_ACCOUNT_BALANCE_HISTORY = exports.GET_FIAT_RATES_FOR_TIMESTAMPS = exports.GET_CURRENT_FIAT_RATES = exports.GET_INFO = exports.DISCONNECT = exports.CONNECT = exports.INIT = exports.HANDSHAKE = void 0;
exports.HANDSHAKE = 'm_handshake';
exports.INIT = 'm_init';
exports.CONNECT = 'm_connect';
exports.DISCONNECT = 'm_disconnect';
exports.GET_INFO = 'm_get_info';
exports.GET_CURRENT_FIAT_RATES = 'm_get_current_fiat_rates';
exports.GET_FIAT_RATES_FOR_TIMESTAMPS = 'm_get_fiat_rates_for_timestamps';
exports.GET_ACCOUNT_BALANCE_HISTORY = 'm_get_account_balance_history';
exports.GET_FIAT_RATES_TICKERS_LIST = 'm_get_fiat_rates_tickers_list';
exports.GET_BLOCK_HASH = 'm_get_block_hash';
exports.GET_BLOCK = 'm_get_block';
exports.GET_ACCOUNT_INFO = 'm_get_account_info';
exports.GET_ACCOUNT_UTXO = 'm_get_account_utxo';
exports.GET_TRANSACTION = 'm_get_transaction';
exports.GET_TRANSACTION_HEX = 'm_get_transaction_hex';
exports.ESTIMATE_FEE = 'm_estimate_fee';
exports.RPC_CALL = 'm_rpc_call';
exports.SUBSCRIBE = 'm_subscribe';
exports.UNSUBSCRIBE = 'm_unsubscribe';
exports.PUSH_TRANSACTION = 'm_push_tx';
exports.TERMINATE = 'terminate';
