// @trezor/blockchain-link-types/lib/constants/responses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NOTIFICATION = exports.DISCONNECTED = exports.CONNECTED = exports.PUSH_TRANSACTION = exports.UNSUBSCRIBE = exports.SUBSCRIBE = exports.RPC_CALL = exports.ESTIMATE_FEE = exports.GET_TRANSACTION_HEX = exports.GET_TRANSACTION = exports.GET_ACCOUNT_BALANCE_HISTORY = exports.GET_ACCOUNT_UTXO = exports.GET_ACCOUNT_INFO = exports.GET_FIAT_RATES_TICKERS_LIST = exports.GET_FIAT_RATES_FOR_TIMESTAMPS = exports.GET_CURRENT_FIAT_RATES = exports.GET_BLOCK = exports.GET_BLOCK_HASH = exports.GET_INFO = exports.CONNECT = exports.ERROR = void 0;
exports.ERROR = 'r_error';
exports.CONNECT = 'r_connect';
exports.GET_INFO = 'r_info';
exports.GET_BLOCK_HASH = 'r_get_block_hash';
exports.GET_BLOCK = 'r_get_block';
exports.GET_CURRENT_FIAT_RATES = 'r_get_current_fiat_rates';
exports.GET_FIAT_RATES_FOR_TIMESTAMPS = 'r_get_fiat_rates_for_timestamps';
exports.GET_FIAT_RATES_TICKERS_LIST = 'r_GET_FIAT_RATES_TICKERS_LIST';
exports.GET_ACCOUNT_INFO = 'r_account_info';
exports.GET_ACCOUNT_UTXO = 'r_get_account_utxo';
exports.GET_ACCOUNT_BALANCE_HISTORY = 'r_get_account_balance_history';
exports.GET_TRANSACTION = 'r_get_transaction';
exports.GET_TRANSACTION_HEX = 'r_get_transaction_hex';
exports.ESTIMATE_FEE = 'r_estimate_fee';
exports.RPC_CALL = 'r_rpc_call';
exports.SUBSCRIBE = 'r_subscribe';
exports.UNSUBSCRIBE = 'r_unsubscribe';
exports.PUSH_TRANSACTION = 'r_push_tx';
exports.CONNECTED = 'r_connected';
exports.DISCONNECTED = 'r_disconnected';
exports.NOTIFICATION = 'r_notification';
