// @trezor/protobuf/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceModelInternal = exports.MessagesSchema = exports.createMessageFromType = exports.createMessageFromName = exports.parseConfigure = exports.Messages = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./decode"), exports);
tslib_1.__exportStar(require("./encode"), exports);
exports.Messages = tslib_1.__importStar(require("./messages"));
tslib_1.__exportStar(require("./types"), exports);
var utils_1 = require("./utils");
Object.defineProperty(exports, "parseConfigure", { enumerable: true, get: function () { return utils_1.parseConfigure; } });
Object.defineProperty(exports, "createMessageFromName", { enumerable: true, get: function () { return utils_1.createMessageFromName; } });
Object.defineProperty(exports, "createMessageFromType", { enumerable: true, get: function () { return utils_1.createMessageFromType; } });
exports.MessagesSchema = tslib_1.__importStar(require("./messages-schema"));
var messages_schema_1 = require("./messages-schema");
Object.defineProperty(exports, "DeviceModelInternal", { enumerable: true, get: function () { return messages_schema_1.DeviceModelInternal; } });
